File: TranslationFiestaCSharp/BatchProcessor.cs
using System;
using System.IO;
using System.Threading.Tasks;

namespace TranslationFiestaCSharp
{
    public class BatchProcessor
    {
        private readonly TranslationClient _translationClient;
        private readonly Action<int, int> _updateCallback;
        private bool _isRunning;

        public BatchProcessor(TranslationClient translationClient, Action<int, int> updateCallback)
        {
            _translationClient = translationClient;
            _updateCallback = updateCallback;
        }

        public async Task ProcessDirectoryAsync(string directoryPath)
        {
            _isRunning = true;
            var files = Directory.GetFiles(directoryPath, "*.*", SearchOption.AllDirectories);
            var textFiles = Array.FindAll(files, f => f.EndsWith(".txt") || f.EndsWith(".md") || f.EndsWith(".html"));
            int totalFiles = textFiles.Length;

            Logger.Info($"Starting batch processing for {totalFiles} files in {directoryPath}");

            for (int i = 0; i < totalFiles; i++)
            {
                if (!_isRunning) break;

                var file = textFiles[i];
                Logger.Info($"Processing file {i + 1}/{totalFiles}: {Path.GetFileName(file)}");

                try
                {
                    string originalContent = await File.ReadAllTextAsync(file);

                    // Perform back-translation: English -> Japanese -> English
                    string japaneseContent = await _translationClient.TranslateAsync(originalContent, "en", "ja");
                    string backtranslatedContent = await _translationClient.TranslateAsync(japaneseContent, "ja", "en");

                    await SaveTranslatedFileAsync(file, backtranslatedContent);

                }
                catch (Exception ex)
                {
                    Logger.Error($"Failed to process file {Path.GetFileName(file)}", ex);
                }

                _updateCallback?.Invoke(i + 1, totalFiles);
            }

            _isRunning = false;
            Logger.Info("Batch processing completed.");
        }

        private async Task SaveTranslatedFileAsync(string originalFilePath, string translatedContent)
        {
            string directory = Path.GetDirectoryName(originalFilePath) ?? "";
            string fileNameWithoutExt = Path.GetFileNameWithoutExtension(originalFilePath);
            string extension = Path.GetExtension(originalFilePath);
            string newFileName = $"{fileNameWithoutExt}_translated{extension}";
            string newFilePath = Path.Combine(directory, newFileName);

            try
            {
                using (var writer = new StreamWriter(newFilePath, false, System.Text.Encoding.UTF8))
                {
                    await writer.WriteLineAsync(translatedContent);
                }

                Logger.Info($"Saved translated file to {newFilePath}");
            }
            catch (Exception ex)
            {
                Logger.Error($"Failed to save translated file {newFilePath}", ex);
            }
        }

        public void Stop()
        {
            _isRunning = false;
        }
    }
}

File: TranslationFiestaFSharp/BatchProcessor.fs
/// <summary>
/// This module provides functionality for batch processing of translation tasks across multiple files in a directory.
/// It has been refactored according to Clean Code principles to improve readability,
/// maintainability, and efficiency, focusing on clear responsibilities and robust error handling.
///
/// Key Refactorings:
/// - **Meaningful Naming**: Ensured functions and parameters have clear, descriptive names.
/// - **Clear Structure**: The module is focused solely on iterating through files and applying a translation function.
/// - **Robust Error Handling**: Replaced direct `MessageBox.Show` calls with a functional `Result` type
///   for error propagation, allowing the calling context (e.g., UI layer) to handle error display.
/// - **Immutability**: Maintained an immutable approach to file processing where possible.
/// - **XML Documentation**: Added comprehensive XML documentation to the module and its public functions.
/// </summary>
namespace TranslationFiestaFSharp

open System
open System.IO
open System.Threading.Tasks // Still needed for Async.AwaitTask if used within translateFn
open System.Windows.Forms

/// <summary>
/// This module provides functionality for batch processing of files within a selected directory.
/// </summary>
module BatchProcessor =
    
    /// <summary>
    /// Processes multiple files in a selected directory using a provided translation function.
    /// </summary>
    /// <param name="translateFn">An asynchronous function that takes a file path and returns
    /// an <see cref="Async{Result{string,string}}"/> representing the translated text or an error.</param>
    /// <param name="updateCallback">A callback function to report progress (completed files, total files).</param>
    /// <returns>An asynchronous operation that, when completed, indicates the result of the batch process for each file.</returns>
    let processDirectory (translateFn: string -> Async<Result<string, string>>) (updateCallback: int -> int -> unit) : Async<Result<string, string> list> =
        async {
            let fbd = new FolderBrowserDialog()
            fbd.Description <- "Select a directory to batch process for translation."

            if fbd.ShowDialog() = System.Windows.Forms.DialogResult.OK then
                let selectedDirectory = fbd.SelectedPath
                let supportedExtensions = [".txt"; ".md"; ".html"; ".epub"]

                let filesToProcess =
                    Directory.GetFiles(selectedDirectory, "*.*", SearchOption.AllDirectories)
                    |> Array.filter (fun f -> supportedExtensions |> List.exists (fun ext -> f.ToLowerInvariant().EndsWith(ext)))
                    |> Array.toList
                
                let totalFiles = filesToProcess.Length
                let mutable completedFiles = 0
                let mutable results = []

                for file in filesToProcess do
                    let! translationResult = translateFn file

                    let processResult =
                        match translationResult with
                        | Ok translatedText ->
                            match Path.GetDirectoryName(file) with
                            | null -> Error (sprintf "Invalid file path for %s: could not determine directory" file)
                            | directory ->
                                let newFileName = Path.Combine(directory, $"{Path.GetFileNameWithoutExtension(file)}_translated{Path.GetExtension(file)}")
                                try
                                    File.WriteAllText(newFileName, translatedText, System.Text.Encoding.UTF8)
                                    Ok (sprintf "Successfully translated %s to %s" file newFileName)
                                with ex ->
                                    Error (sprintf "Failed to write translated file %s: %s" newFileName ex.Message)
                        | Error msg ->
                            Error (sprintf "Translation failed for %s: %s" file msg)
                    
                    results <- processResult :: results
                    
                    completedFiles <- completedFiles + 1
                    updateCallback completedFiles totalFiles
                
                return List.rev results
            else
                return []
        }
